src/apps/replication/lib/mutation_log.cpp

```diff
@@ -1278,8 +1278,149 @@ int mutation_log::garbage_collection(replica_log_info_map& gc_condition)
+                //We hope that this never happens, it would deteriorate performance
+                if (size > writer.total_size())
+                {
+                    auto task = file::read(file_handle, writer.get_current_buffer().buffer().get() + writer.total_size(), size - writer.total_size(), file_dispatched_bytes, LPC_AIO_IMMEDIATE_CALLBACK, nullptr, nullptr);
+                    task->wait();
+                    writer.write_empty(task->io_size());
+                    file_dispatched_bytes += task->io_size();
+                    TRY(task->error());
+                }
+            }
+            result = writer.get_current_buffer();
+        }
+        fill_buffers();
+        return ERR_OK;
+#undef TRY
+    }
+private:
+    void fill_buffers()
+    {
+        while (!current_buffer->have_ongoing_task && current_buffer->empty())
+        {
+            current_buffer->begin = current_buffer->end = 0;
+            current_buffer->file_offset_of_buffer = file_dispatched_bytes;
+            current_buffer->have_ongoing_task = true;
+            current_buffer->task = file::read(file_handle, current_buffer->buffer.get(), block_size_bytes, file_dispatched_bytes, LPC_AIO_IMMEDIATE_CALLBACK, nullptr, nullptr);
+            file_dispatched_bytes += block_size_bytes;
+            std::swap(current_buffer, next_buffer);
+        }
+    }
+
+    //buffer size, in bytes
+    static const size_t block_size_bytes = 1024 * 1024;
+    struct buffer_t
+    {
+        std::unique_ptr<char[]> buffer; //with block_size
+        size_t begin, end; // [buffer[begin]..buffer[end]) contains unconsumed_data
+        size_t file_offset_of_buffer; //file offset projected to buffer[0]
+        bool have_ongoing_task;
+        task_ptr task;
+        buffer_t() : buffer(new char[block_size_bytes]), begin(0), end(0), have_ongoing_task(false), file_offset_of_buffer(0) {}
+        size_t length() const
+        {
+            return end - begin;
+        }
+        bool empty() const
+        {
+            return length() == 0;
+        }
+        void consume(binary_writer& dest, size_t len)
+        {
+            dest.write(buffer.get() + begin, len);
+            begin += len;
+        }
+        size_t drain(binary_writer& dest)
+        {
+            auto len = length();
+            consume(dest, len);
+            return len;
+        }
+        error_code wait_ongoing_task()
+        {
+            if (have_ongoing_task)
+            {
+                task->wait();
+                have_ongoing_task = false;
+                end += task->io_size();
+                dassert(end <= block_size_bytes, "invalid io_size.");
+                return task->error();
+            }
+            else
+            {
+                return ERR_OK;
+            }
+        }
+    } buffers[2];
+    buffer_t* current_buffer, *next_buffer;
+
+    //number of bytes we have issued read operations
+    size_t file_dispatched_bytes;
```

Comment:
use underscore as prefix for those members to differentiate with the local variables